core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecider.java

```diff
@@ -98,4 +98,15 @@ public Decision canAllocate(RoutingNode node, RoutingAllocation allocation) {
     public Decision canRebalance(RoutingAllocation allocation) {
         return Decision.ALWAYS;
     }
+
+    /**
+     * Returns a {@link Decision} whether the given primary shard can be
+     * forcibly allocated on the given node. This method should only be called
+     * for unassigned primary shards where the node has a shard copy on disk.
+     */
+    public Decision canForceAllocatePrimary(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        assert shardRouting.primary() : "must not call canForceAllocatePrimary on a non-primary shard " + shardRouting;
+        assert shardRouting.unassigned() : "must not call canForceAllocatePrimary on an assigned shard " + shardRouting;
+        return Decision.YES; // by default, a decider will allow force allocation of the primary
```

Comment:
I think I prefer the way it was before with the default being 

```
decision = canAllocate(...)
if (decision == Decision.NO) {
    decision = Decision.single(Type.YES, "force override of " + decision.label, ...)
}
```

This keeps the override logic out of `AllocationDeciders`.